File: diagram-poc/backend/src/main/java/com/example/diagram/service/VersionService.java
package com.example.diagram.service;

import com.example.diagram.web.dto.VersionDetail;
import com.example.diagram.web.dto.VersionRequest;
import com.example.diagram.web.dto.VersionSummary;

import java.util.List;

/** Version history (snapshots) for diagrams. */
public interface VersionService {

    VersionSummary snapshot(Long diagramId, VersionRequest request, String userEmail);

    List<VersionSummary> list(Long diagramId);

    VersionDetail get(Long versionId);
}

File: diagram-poc/frontend/src/app/core/services/version.service.ts
import { Injectable } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { Observable } from 'rxjs';
import { apiBaseUrl } from '../app-config';

export interface VersionSummary {
  id: number;
  label: string;
  authorName: string;
  createdAt: string;
}

export interface VersionDetail extends VersionSummary {
  contentJson: string;
}

const API = apiBaseUrl();

/** Version history (snapshots) for a saved diagram. */
@Injectable({ providedIn: 'root' })
export class VersionService {
  constructor(private http: HttpClient) {}

  list(diagramId: number): Observable<VersionSummary[]> {
    return this.http.get<VersionSummary[]>(`${API}/diagrams/${diagramId}/versions`);
  }

  snapshot(diagramId: number, label: string, contentJson: string): Observable<VersionSummary> {
    return this.http.post<VersionSummary>(`${API}/diagrams/${diagramId}/versions`, { label, contentJson });
  }

  get(versionId: number): Observable<VersionDetail> {
    return this.http.get<VersionDetail>(`${API}/versions/${versionId}`);
  }
}
